src/sensors.cpp

```diff
@@ -1,261 +1,201 @@
-volatile uint64_t n3_total = 0;
-volatile uint64_t n2_avgs[RPM_SAMPLES_DEBOUNCE] = {0,0,0,0,0,0,0,0,0,0};
-volatile uint64_t n2_total = 0;
-volatile uint64_t output_avgs[RPM_SAMPLES_DEBOUNCE] = {0,0,0,0,0,0,0,0,0,0};
-volatile uint64_t output_total = 0;
+uint32_t n3_avgs[RPM_SAMPLES_DEBOUNCE] = {0,0,0,0,0,0,0,0,0,0};
+volatile uint32_t n3_total = 0;
+uint32_t n2_avgs[RPM_SAMPLES_DEBOUNCE] = {0,0,0,0,0,0,0,0,0,0};
+volatile uint32_t n2_total = 0;
+uint32_t output_avgs[RPM_SAMPLES_DEBOUNCE] = {0,0,0,0,0,0,0,0,0,0};
+volatile uint32_t output_total = 0;
 uint8_t avg_n2_idx = 0;
 uint8_t avg_n3_idx = 0;
 uint8_t avg_out_idx = 0;
-
-
-uint64_t t_n2 = 0;
-uint64_t t_n3 = 0;
-uint64_t t_output = 0;
 bool output_rpm_ok = false;
 
-static void IRAM_ATTR on_pcnt_overflow_n2(void *args)
-{
-    t_n2 = esp_timer_get_time();
-    if (t_n2 - n2_intr_times[1] > 10)
-    {
-        portENTER_CRITICAL_ISR(&n2_mux);
-        n2_intr_times[0] = n2_intr_times[1];
-        n2_intr_times[1] = t_n2;
-        portEXIT_CRITICAL_ISR(&n2_mux);
-    }
-    
-}
 
-static void IRAM_ATTR on_pcnt_overflow_n3(void *args)
-{
-    t_n3 = esp_timer_get_time();
-    if (t_n3 - n3_intr_times[1] > 10)
-    {
-        portENTER_CRITICAL_ISR(&n3_mux);
-        n3_intr_times[0] = n3_intr_times[1];
-        n3_intr_times[1] = t_n3;
-        portEXIT_CRITICAL_ISR(&n3_mux);
-    }
-}
-
-static void IRAM_ATTR on_pcnt_overflow_output(void* args) {
-    t_output = esp_timer_get_time();
-    if (t_output - output_intr_times[1] > 10) {
-        portENTER_CRITICAL_ISR(&output_mux);
-        output_intr_times[0] = output_intr_times[1];
-        output_intr_times[1] = t_output;
-        portEXIT_CRITICAL_ISR(&output_mux);
+static esp_err_t IRAM_ATTR read_and_reset_pcnt(pcnt_unit_handle_t unit, int* dest) {
+    esp_err_t res = pcnt_unit_get_count(unit, dest);
+    if (ESP_OK == res) {
+        pcnt_unit_clear_count(unit);
     }
+    return res;
 }
 
-static intr_handle_t rpm_timer_handle;
-static void IRAM_ATTR on_rpm_timer(void* args) {
-    // Clear timer interrupt
-    TIMERG0.int_clr_timers.t0 = 1;
-    TIMERG0.hw_timer[0].config.alarm_en = 1;
-    uint64_t time_delta_n2 = 0;
-    uint64_t time_delta_n3 = 0;
-    uint64_t time_delta_output = 0;
-    portENTER_CRITICAL_ISR(&n2_mux);
-    time_delta_n2 = n2_intr_times[1] - n2_intr_times[0];
-    portEXIT_CRITICAL_ISR(&n2_mux);
-    portENTER_CRITICAL_ISR(&n3_mux);
-    time_delta_n3 = n3_intr_times[1] - n3_intr_times[0];
-    portEXIT_CRITICAL_ISR(&n3_mux);
-    portENTER_CRITICAL_ISR(&output_mux);
-    time_delta_output = output_intr_times[1] - output_intr_times[0];
-    portEXIT_CRITICAL_ISR(&output_mux);
+static bool IRAM_ATTR on_rpm_timer(gptimer_handle_t timer, const gptimer_alarm_event_data_t *edata, void *user_data) {
+    int pulses = 0;
+    // N2 Sensor
+    if (ESP_OK == read_and_reset_pcnt(PCNT_HANDLE_N2, &pulses)) {
+        n2_total = n2_total - n2_avgs[avg_n2_idx];
```

Comment:
I would suggest to write "n2_total -= n2_avgs[avg_n2_idx];". It's in general safer von multi threading environments and avoids to use a wrong variable.

Reply:
actually no. with n2_total (And similar) being volatile (Due to syncing between cores depending on which core handles the interrupt), -= assignment is deprecated in C++20 with volatile. So this is the only way to do it without making the compiler complain.